finesse/hardware/device.py

```diff
@@ -0,0 +1,171 @@
+The DeviceBaseType is a base class for *types* of devices (e.g. a stepper motor). All
+concrete classes for devices must inherit from a particular DeviceBaseType.
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from typing import Any
+
+from finesse.device_info import DeviceBaseTypeInfo, DeviceParameter, DeviceTypeInfo
+
+from .plugins import load_all_plugins
+
+_base_types: set[type[DeviceBaseType]] = set()
+"""Registry of device base types."""
+
+_device_types: set[type[Device]] = set()
+"""Registry of concrete device types."""
+
+
+def get_device_type_registry() -> dict[DeviceBaseTypeInfo, list[DeviceTypeInfo]]:
+    """Return info about device types grouped according to their base type."""
+    # Ensure all base types and device types have been registered
+    load_all_plugins()
+
+    # Get the base type info and sort it alphabetically by description
+    base_types_info = sorted(
+        (t.get_device_base_type_info() for t in _base_types),
+        key=lambda info: info.description,
+    )
+
+    # Preallocate dict with empty lists
+    out: dict[DeviceBaseTypeInfo, list[DeviceTypeInfo]] = {
+        info: [] for info in base_types_info
+    }
+
+    # Get device type info and group by base type
+    for device_type in _device_types:
+        out[device_type.get_device_base_type_info()].append(
+            device_type.get_device_type_info()
+        )
+
+    # Sort the device types by name
+    for infos in out.values():
+        infos.sort(key=lambda info: info.description)
+
+    return out
+
+
+class Device(ABC):
+    """A base class for all devices."""
+
+    _device_base_type_info: DeviceBaseTypeInfo
+    """Information about the device's base type."""
+    _device_description: str
+    """A human-readable name."""
+    _device_parameters: list[DeviceParameter] = []
+    """Possible parameters that this device type accepts.
+
+    The key represents the parameter name and the value is a list of possible values.
+    """
+
+    def __init_subclass__(cls, description: str | None = None, **kwargs: Any) -> None:
+        """Initialise a device type class.
+
+        While the description argument has to be optional to allow for non-concrete
+        classes to inherit from Device, it is mandatory to include it, else the device
+        type class will not be added to the registry.
+
+        Args:
+            description: Human-readable name for this device type.
+        """
+        # Forward keyword args to allow for multiple inheritance
+        super().__init_subclass__(**kwargs)
+
+        # **HACK**: Allow callers to omit this arg in the case that they are
+        # non-concrete
+        if not description:
+            return
+
+        # Set device description for this class
+        cls._device_description = description
+
+        # Add the class to the registry of device types
+        _device_types.add(cls)
```

Comment:
I think I might be missing something. Is there any reason why you cannot do the following and skip the `DeviceBaseType` class below altogether?


```suggestion
    def __init_subclass__(cls, is_base_type: bool = False, **kwargs: Any) -> None:
        """Initialise a device type class.

        While the description argument has to be optional to allow for non-concrete
        classes to inherit from Device, it is mandatory to include it, else the device
        type class will not be added to the registry.

        Args:
            description: Human-readable name for this device type.
        """
        # If it is a base class, we initialise it as such
        if is_base_type:
            cls._init_base_type(**kwargs)
            return

        # If it is not, it should inherit from one.
        # Not sure how robust this will be to differences in the registration order...
        if not set(cls.__bases__).intersection(_base_types):
            raise ValueError(f"Class {cls.__name__} must be a device base type or inherit from one.")

        # And we initialise it as a concrete device class
        cls._init_device_type(**kwargs)

    @classmethod
    def _init_base_type(
        cls,
        name: str,
        description: str,
        names_short: Sequence[str] = (),
        names_long: Sequence[str] = (),
        **kwargs,
    ) -> None:
        super().__init_subclass__(**kwargs)
        # Store metadata about this base class
        cls._device_base_type_info = DeviceBaseTypeInfo(
            name, description, names_short, names_long
        )

        # Add the class to the registry of base types
        _base_types.add(cls)

    @classmethod
    def _init_device_type(
        cls,
        name: str,
        description: str,
        **kwargs,
    ) -> None:
        super().__init_subclass__(**kwargs)
        # Set device description for this class
        cls._device_description = description

        # Add the class to the registry of device types
        _device_types.add(cls)
```

Reply:
You can even go one step further and take the decision of a class being base or not depending on the class name instead of an input parameter (although that's more explicit): if it ends in `Base` it is a base class and not otherwise. `StepMotorBase` will be considered a base class, so will `TempControllerBase`, but not `TempController` or `MyFunnyClassWithOddName`.

Reply:
Interesting! I like the idea of specifying `is_base_type` explicitly. I'll have a tinker
with this.

The reason for having a separate `DeviceBaseType` class was partly because
`SerialDevice` inherits from `Device` but *not* `DeviceBaseType`, because it isn't a
concrete device implementation (it isn't a device base type either actually). I guess we
could either a) make an `AbstractDevice` parent class which includes just the bits actually
shared with `SerialDevice` or b) make `SerialDevice` not inherit from anything (not sure
if this will break something or not).

Reply:
So, how this goes? I think I'm missing the overall architecture of classes you are implementing. 

In the current implementation, I thought you would have a, lets say, `StepperMotorBase` inheriting from `DeviceBaseType`, and a `ModelXYZStepperMotor` as a concrete, usable device, inheriting from `StepperMotorBase` (because it is a stepper motor) and `SerialDevice` (because it is a serial device). Why does `SerialDevice` need to inherit from `Device` without being a `DeviceBaseType`? Or why it needs to inherit from `Device` at all?

It seems to me you might be mixing two orthogonal concepts here: device types (stepper motor, temp controller, etc.) with connection types (serial, GPIB, ethernet...). It makes sense to me that a concrete device needs to inherit from both (or use composition, instead), but I don't see any reason why these two concepts need to have a common ancestor to inherit from.

Reply:
Right, so I've finally finished doing this.

I did end up making an `AbstractDevice` class as a separate base class. The reason is that the `SerialDevice` class needs to modify a class attribute (device parameters) which is shared with other device types.